AJ-581: Add arrays as data type

sagehen03 opened the pull request:
This PR adds the ability to upload a JSON representation of an array and store that data in a postgres array column.  Boolean, String, Number, LocalDate, and LocalDateTime are the supported array types. For the date types it was necessary to convert to String[] and then cast in the SQL insert in order for db writes to succeed. Handling both the TSV and the JSON representations of the data increases the complexity.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/WorkspaceDataServiceApplication.java:
```diff
@@ -27,8 +32,19 @@ public class WorkspaceDataServiceApplication {
+				.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+				.configure(DeserializationFeature.ACCEPT_FLOAT_AS_INT, false)
+				.findAndAddModules().build();
+		mapper.coercionConfigFor(LogicalType.Boolean).setCoercion(CoercionInputShape.Integer, CoercionAction.Fail);
```
Don't coerce 0 to false and everything else to true.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/WorkspaceDataServiceApplication.java:
```diff
@@ -27,8 +32,19 @@ public class WorkspaceDataServiceApplication {
+				.configure(DeserializationFeature.ACCEPT_FLOAT_AS_INT, false)
+				.findAndAddModules().build();
+		mapper.coercionConfigFor(LogicalType.Boolean).setCoercion(CoercionInputShape.Integer, CoercionAction.Fail);
+		mapper.coercionConfigFor(LogicalType.Float).setCoercion(CoercionInputShape.String, CoercionAction.Fail);
+		mapper.coercionConfigFor(LogicalType.Integer).setCoercion(CoercionInputShape.String, CoercionAction.Fail);
```
Quoted numbers should stay strings, aka don't make ["1", "2"] an array of long.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/WorkspaceDataServiceApplication.java:
```diff
@@ -27,8 +32,19 @@ public class WorkspaceDataServiceApplication {
-				.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false).findAndAddModules().build();
+		JsonMapper mapper = JsonMapper.builder().enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
+				.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+				.configure(DeserializationFeature.ACCEPT_FLOAT_AS_INT, false)
```
Don't convert floats to ints.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -127,7 +128,7 @@ public Map<String, DataTypeMapping> getExistingTableSchema(UUID instanceId, Reco
 		params.addValue("recordName", RECORD_ID);
 		return namedTemplate
-				.query("select column_name, data_type from INFORMATION_SCHEMA.COLUMNS where table_schema = :instanceId "
+				.query("select column_name, udt_name::regtype as data_type from INFORMATION_SCHEMA.COLUMNS where table_schema = :instanceId "
```
This change was necessary to get the array type aka text[] as opposed to just returning ARRAY.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -368,9 +369,43 @@ private Object getValueForSql(Object attVal, DataTypeMapping typeMapping) {
+			case ARRAY_OF_BOOLEAN:
+				return attVal.toArray(new Boolean[0]);
+			case ARRAY_OF_DOUBLE:
+				List<Object> doubleList = attVal.stream().map(i -> {
+					if (i instanceof Integer){
+						return (double) ((Integer) i).intValue();
+					}
+					else {
+						return i;
+					}
+				}).toList();
```
I hate this, but I haven't been able to find a better way.  If Spring serializes from JSON something like `attrFoo: [1, 2.2]`, that will reach this part of the code as an ArrayList containing an Integer and Double and Java won't let us just call `attVal.toArray(new Double[0])` https://stackoverflow.com/questions/5566336/classcastexception-casting-integer-to-double

sagehen03 replied:
It's also worth mentioning that arrays with very large numbers aren't supported.  The json serialization will convert those numbers to BigDecimal of BigInteger, but postgres' supported numeric array types can't handle numbers that size.  I think the best we can do in that case is store the data as text or json.  It also looks like very big numbers for plain (non-array) numeric attributes will overflow which is not unexpected since we're coercing to Double and Long.  All of this opens up the question of what scale of numeric values do we need to support?

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -82,6 +96,8 @@ public DataTypeMapping inferTypeForTsvSource(Object val) {
 			return STRING;
 		}
+		//libreoffice at least uses left and right quotes which cause problems when we try to parse as JSON
+		sVal = sVal.replaceAll("[“”]", "\"");
```
When I created a TSV in my spreadsheet program it used left and right quotes, which made Jackson puke.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/service/model/DataTypeMapping.java:
```diff
@@ -2,33 +2,41 @@
-		MAPPING_BY_PG_TYPE.put("bigint", LONG);
-		MAPPING_BY_PG_TYPE.put("numeric", DOUBLE);
-		MAPPING_BY_PG_TYPE.put("boolean", BOOLEAN);
+		Arrays.stream(DataTypeMapping.values()).forEach(e -> MAPPING_BY_PG_TYPE.put(e.getPostgresType(), e));
```
Oops I realized there was a more concise way to do this and keep things DRYer

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/WorkspaceDataServiceApplication.java:
```diff
@@ -26,9 +30,18 @@ public class WorkspaceDataServiceApplication {
+	}
+
+	@Bean
+	public DataTypeInferer inferer(ObjectMapper mapper){
```
We've changed a number of defaults on the Object mapper, there are a few changes like this to share a single instance across the app so that we get consistent JSON (de)serialization behavior.

calypsomatic commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -57,6 +64,12 @@ public DataTypeMapping selectBestType(DataTypeMapping existing, DataTypeMapping
+			return newMapping != EMPTY_ARRAY ? newMapping : existing;
+		}
+		if(newMapping.isArrayType() && existing.isArrayType()){
+			return ARRAY_OF_STRING;
```
I'm having a hard time following where something is for sql or tsv or whatnot - so this looks like `ARRAY_OF_STRING` is the only array type we have, so is this method just for the database?

calypsomatic replied:
Aha!  Gotcha.  Thanks

calypsomatic commented on service/src/test/resources/small-test.tsv:
```diff
@@ -1,5 +1,5 @@
-sys_name	greeting	unicode	json	date	embedded characters	double	long
+sys_name	greeting	unicode	json	date	embedded characters	double	long	z_array_of_string	z_long_array	z_double_array	z_z_boolean_array	zz_array_of_date	zz_array_of_datetime
```
Are these z's just to keep it alphabetical order?

sagehen03 replied:
yes!

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -469,6 +497,13 @@ private RecordAttributes getAttributes(ResultSet rs) {
+							if(object.getClass() == Timestamp[].class) {
+								object = convertToLocalDateTime(object);
+							} else if(object.getClass() == Date[].class) {
+								object = convertToLocalDate(object);
```
I suggest using `object instanceof Date[] dtArray` to cast the object up here, instead of inside `convertToLocalDate` - and changing `convertToLocalDate` to accept a `Date[]` instead of an `Object`. This is minor, but it prevents a future coder from a bad call to `convertToLocalDate` and then running into a ClassCastException. Same for `convertToLocalDateTime`.

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -479,6 +514,24 @@ private RecordAttributes getAttributes(ResultSet rs) {
+			for (int i = 0; i < tzArray.length; i++) {
+				result[i] = tzArray[i].toLocalDate();
+			}
+			return result;
```
both `convertToLocalDateTime` and `convertToLocalDate` could use streams, e.g.
```java
return Arrays.stream(tzArray).map(Date::toLocalDate).toArray(LocalDate[]::new);
```

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -469,6 +497,13 @@ private RecordAttributes getAttributes(ResultSet rs) {
+								object = convertToLocalDateTime(object);
+							} else if(object.getClass() == Date[].class) {
+								object = convertToLocalDate(object);
+							}
```
also - are there unit tests for these date/timestamp arrays? I was messing around with the code here and couldn't get tests to fail, no matter what I changed.

sagehen03 replied:
What did you change?  There are definitely tests that write to these array date types, but I think we could add some tests for reading and verifying the proper value.  I will do add those.

davidangb replied:
this still passes tests:
```java
						} else if (object instanceof PgArray pgArray) {
							object = pgArray.getArray();
							if(object.getClass() == Timestamp[].class) {
								object = "failme";
							} else if(object.getClass() == Date[].class) {
								object = 1234;
							}
						}
```

sagehen03 replied:
Ah that gibes with my understanding of the tests.  I was trying to keep this PR from ever expanding and was going to add a follow on ticket to clean up the tests as we discussed on Tuesday and include in that more validation that we're returning the proper values, but I'll add some date read validation now.

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -57,6 +64,12 @@ public DataTypeMapping selectBestType(DataTypeMapping existing, DataTypeMapping
+		}
+		if(newMapping.isArrayType() && existing.isArrayType()){
+			return ARRAY_OF_STRING;
+		}
```
I think we should discuss requirements with PM first before making any additional changes here … but for future thought: if existing is DATE and newMapping is DATETIME, we could change to DATETIME instead of STRING

sagehen03 replied:
Good call!

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -84,7 +97,12 @@ public DataTypeMapping inferTypeForTsvSource(Object val) {
+
+	//libreoffice at least uses left and right quotes which cause problems when we try to parse as JSON
+	private String replaceLeftRightQuotes(String val){
+		return val.replaceAll("[“”]", "\"");
```
my opinion is we shouldn't do this at all - but if we do, we should only replace leading/trailing smart quotes; we shouldn't replace anything in the middle of the string

sagehen03 replied:
I will test and see what jackson does with non-leading left right quotes.  If it doesn't puke, I will adjust the regex to look only for leading and trailing quotes.

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -162,13 +186,64 @@ public boolean isNumericValue(String sVal) {
+			}
+		}
+		return DataTypeMapping.getArrayTypeForBase(bestMapping);
+	}
```
what I was picturing when we spoke in person was something like this:
```java
	private <T> DataTypeMapping findArrayType(List<T> list){
		DataTypeMapping bestMapping = null;
		// find the distinct inferred types from the input
		List<DataTypeMapping> inferredTypes = list.stream()
				.map(item -> inferType(item, InBoundDataSource.JSON))
				.distinct()
				.toList();
		if (inferredTypes.size() == 1) {
			bestMapping = inferredTypes.get(0);
		} else {
			for (DataTypeMapping type : inferredTypes) {
				bestMapping = selectBestType(bestMapping, type);
			}
		}
		return DataTypeMapping.getArrayTypeForBase(bestMapping);
	}
```

sagehen03 replied:
That seems reasonable.

davidangb commented on service/src/main/java/org/databiosphere/workspacedataservice/service/DataTypeInferer.java:
```diff
@@ -162,13 +186,64 @@ public boolean isNumericValue(String sVal) {
+			return objectMapper.readValue(replaceLeftRightQuotes(val), clazz);
 		} catch (JsonProcessingException e) {
-			return false;
+			return null;
```
thinking out loud … 

`findArrayTypeFromJson` and `getArrayOfType` seem pretty heavyweight, but I can't realistically think of anything better to do ... maybe we just keep an eye here for any potential problems?

sagehen03 replied:
That's usually my preference!

davidangb reviewed (commented):
see inline! This is a big complex PR

sagehen03 commented on service/src/main/resources/static/swagger/openapi-docs.yaml:
```diff
@@ -518,7 +518,7 @@ components:
         datatype:
           type: string
-          enum: [boolean, date, datetime, string, json, long, double, relation]
+          enum: [boolean, date, datetime, string, json, long, double, relation, array_of_boolean, array_of_string, array_of_number, array_of_date, array_of_datetime]
```
Your suggestion feels more complicated to me.

sagehen03 commented on service/src/test/java/org/databiosphere/workspacedataservice/DataTypeInfererTest.java:
```diff
@@ -101,18 +105,34 @@ void inferSomeTypes() {
+		assertThat(inferer.inferType("", InBoundDataSource.JSON)).isEqualTo(DataTypeMapping.STRING);
+		assertThat(inferer.inferType("[11, 99, -3.14]", InBoundDataSource.JSON)).isEqualTo(DataTypeMapping.ARRAY_OF_NUMBER);
+		assertThat(inferer.inferType("[11, 99, -3.14, 09]", InBoundDataSource.JSON)).isEqualTo(DataTypeMapping.STRING);
+		assertThat(inferer.inferType("[]", InBoundDataSource.JSON)).isEqualTo(DataTypeMapping.EMPTY_ARRAY);
```
I agree with you on the inconsistency of quoted coercible boolean values vs. numbers.  I would say if we want to prioritize consistency I would say let's switch to making "true" remain strings.  The existing non array boolean behavior was that we convert them to booleans so I wanted to make array booleans consistent with that, but if we want to re-visit the incosistency between boolean vs. numbers let's do that.  I'll piing Beth and Alessandro.

sagehen03 commented on service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:
```diff
@@ -365,9 +368,33 @@ private Object getValueForSql(Object attVal, DataTypeMapping typeMapping) {
+			return getListAsArray(valAsList, typeMapping);
+		}
+		return inferer.getArrayOfType(attVal.toString(), typeMapping.getJavaArrayTypeForDbWrites());
+	}
```
Correct or if the user submits quoted json as an attribute value via the json code paths.

sagehen03 commented on service/src/test/java/org/databiosphere/workspacedataservice/controller/RecordControllerMockMvcTest.java:
```diff
@@ -131,7 +132,57 @@ void simpleTsvUploadWithBatchingShouldSucceed(@Value("${twds.write.batch.size}")
+			tsvContent.append(i + "valid\t[12]\n");
+		}
+		MockMultipartFile file = new MockMultipartFile("records", "simple.tsv", MediaType.TEXT_PLAIN_VALUE,
+				tsvContent.toString().getBytes());
```
For a narrow tsv with only one or two columns this doesn't seem unwieldy to me.  It's also nice not to have to look at two editor panes to understand the test

sagehen03 commented:
@davidangb I think I've addressed your PR feedback except for two cases: 

string coercion consistency--we're waiting on some guidance in slack I'd like to open a separate ticket for this and address it once we have consensus 

replacing directional quotes--the quotes that trip jackson aren't at the edges of the string so I think it should stay, but if I need to remove them for your approval I can do that and we can see what happens once we start getting usage.

davidangb reviewed (approved):
:+1: thank you!

On the smart-quote issue … could you document what you have run into, in a unit test or in a jira ticket or whatever form makes sense to you? Based on your comment about interior quotes being a problem, I tried a very simple unit test:
```java
	@Test
	void testSmartQuotes() throws JsonProcessingException {
		String input = "\"f“o“o\"";
		String actual = jacksonObjectMapper.readValue(input, String.class);
		assertEquals("f“o“o", actual);
	}
```
and the test passed … so I think I am not understanding the issue correctly and want to make sure we follow up on it.

sagehen03 commented:
> ```java
> String input = "\"f“o“o\"";
> 		String actual = jacksonObjectMapper.readValue(input, String.class);
> 		assertEquals("f“o“o", actual);
> ```
This illustrates the issue with smart quotes.  It might be possible to craft a complicated regex to ferret those element delimiting smart quotes but I feel like complicated regexes never end well.
```
@Test
void quoting() throws JsonProcessingException {
String input = "[“foo”, “bazz”, “bar”]";
String[] actual = mapper.readValue(input, String[].class);
assertArrayEquals(new String[]{"foo", "baz", "bar"}, actual);
}
```
